arcgis-ios-sdk-samples/Layers/Group layers/LayersTableViewController.swift

```diff
@@ -0,0 +1,156 @@
+// You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import UIKit
+import ArcGIS
+
+class LayersTableViewController: UITableViewController, GroupLayersCellDelegate, GroupLayersSectionViewDelegate {
+    var layers = [AGSLayer]()
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        self.tableView.register(GroupLayersSectionView.nib, forHeaderFooterViewReuseIdentifier: GroupLayersSectionView.reuseIdentifier)
+    }
+    
+    // MARK: - UITableViewDataSource
+    
+    override func numberOfSections(in tableView: UITableView) -> Int {
+        return layers.count
+    }
+    
+    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        guard let groupLayer = getLayer(from: layers, at: section) as? AGSGroupLayer else { return 0 }
+        return groupLayer.layers.count
+    }
+    
+    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: "cell", for: indexPath) as? GroupLayersCell else {
+            return UITableViewCell(style: .default, reuseIdentifier: "cell")
+        }
+        
+        guard let groupLayer = getLayer(from: layers, at: indexPath.section) as? AGSGroupLayer, let childLayers = groupLayer.layers as? [AGSLayer], let childLayer = getLayer(from: childLayers, at: indexPath.row) else { return cell }
+        // Set label.
+        cell.textLabel?.text = formattedValue(of: childLayer.name)
+        
+        // Set the switch to on or off.
+        cell.layerVisibilitySwitch.isOn = childLayer.isVisible
+        
+        // To update the visibility of operational layers on switch toggle.
+        cell.delegate = self
+        
+        return cell
+    }
+    
+    // MARK: - UITableViewDelegate
+    
+    override func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard let headerView = tableView.dequeueReusableHeaderFooterView(withIdentifier: GroupLayersSectionView.reuseIdentifier) as? GroupLayersSectionView else { return nil }
+        
+        guard let layer = getLayer(from: layers, at: section) else { return nil }
+        // Set label.
+        headerView.layerNameLabel.text = formattedValue(of: layer.name)
+        
+        // Set the switch to on or off.
+        headerView.layerVisibilitySwitch?.isOn = layer.isVisible
+        if !layer.isVisible {
+            // Disable switch of child layers if the parent is not visible.
+            updateSwitchForRows(in: section, isEnabled: false)
```

Comment:
I'm not sure if I understood what you mean. In `tableView(_:cellForRowAt:)`, I'm setting the switch on or off based on the visibility of child layers. Since child layer's visibility can be true when parent layer's visibility is false, I have "disabled" child layers switch when the parent layer is turned off. That would communicate to user that they can't toggle the child when the parent is off.

Reply:
As far as I can tell, child layer switch values are being set twice: once in `tableView(_:viewForHeaderInSection:)` (by the call to `updateSwitchForRows(in:isEnabled:)`) and again in `tableView(_:cellForRowAt:)`. I don't know which is called first, but I imagine that `tableView(_:viewForHeaderInSection:)` is called before `tableView(_:cellForRowAt:)`. In either case, it would be nice to only set each switch's value once when the table data is loaded. Could `tableView(_:cellForRowAt:)` take into account the visibility of the parent layer when setting the switch value? Perhaps with a helper function:

```swift
func isLayerVisible(at indexPath: IndexPath) -> Bool {
    let groupLayer = layers[indexPath.section] as! AGSGroupLayer
    let childLayer = groupLayer.layers[indexPath.row] as! AGSLayer
    return groupLayer.isVisible && childLayer.isVisible
}
```

and then in `tableView(_:cellForRowAt:)`:

```swift
cell.layerVisibilitySwitch.isOn = isLayerVisible(at: indexPath)
```

Reply:
I see what you mean. But we need to honor child layer's visibility even when the parent is turned off. This is what I mean - 

When group layer visibility is turned off, its child layers won't be visible on map. But those child layer's `isVisible` property could still be `true`. 

`return groupLayer.isVisible && childLayer.isVisible` will turn off the switch of child layer even when the child layer's `isVisible` property is `true`, which is not correct. 

Hence, I am "disabling" the switch of child layers when the parent is turned off. Switch disable is different than swift on/off state. 

I hope I didn't confuse you.

Reply:
Ah, that makes sense. In that case, could we still do that in `tableView(_:cellForRowAt:)`?

```swift
cell.layerVisibilitySwitch.isOn = childLayer.isVisible
cell.layerVisibilitySwitch.isEnabled = layers[indexPath.section].isVisible
```

Reply:
yes, thank you!